source/common/rate_limiter_impl.h

```diff
@@ -54,4 +56,38 @@ class LinearRateLimiter : public RateLimiter,
   Envoy::MonotonicTime started_at_;
 };
 
+using RandomDistributionGenerator =
+    std::function<const std::chrono::duration<uint64_t, std::nano>()>;
+
+// Wraps a rate limiter, and allows plugging in a delegate which will be queried to offset the
+// timing of the underlying rate limiter.
+class RandomDistributingRateLimiter : public RateLimiter,
+                                      public Envoy::Logger::Loggable<Envoy::Logger::Id::main> {
+public:
+  RandomDistributingRateLimiter(Envoy::TimeSource& time_source, RateLimiterPtr&& rate_limiter,
+                                RandomDistributionGenerator random_distribution_generator);
+  bool tryAcquireOne() override;
+  void releaseOne() override;
+
+protected:
+  const RandomDistributionGenerator random_distribution_generator_;
+
+private:
+  Envoy::TimeSource& time_source_;
+  const RateLimiterPtr rate_limiter_;
+  Envoy::MonotonicTime distributed_start_;
+  bool distributed_start_set_{};
+};
+
+// Allows adding uniformly distributed random timing offsets to an underlying rate limiter.
+class UniformDistributingRateLimiter : public RandomDistributingRateLimiter {
+public:
+  UniformDistributingRateLimiter(Envoy::TimeSource& time_source, RateLimiterPtr&& rate_limiter,
+                                 const std::chrono::nanoseconds upper_bound);
+
+private:
+  std::default_random_engine generator_;
+  std::uniform_int_distribution<uint64_t> distribution_;
```

Comment:
Is there a way to do dependency injection of this or control the seed for deterministic tests?

Reply:
Added dc748584f1d4a27739e33dfd2b3ee5e8b6935ce5 which refactors a bit. And rewrote a test to leverage the result.
Let me know if this looks better